sessionpeermanager/sessionpeermanager.go

```diff
@@ -328,7 +328,12 @@ type peerTimeoutMessage struct {
 
 func (ptm *peerTimeoutMessage) handle(spm *SessionPeerManager) {
 	data, ok := spm.activePeers[ptm.p]
-	if !ok || !data.lt.WasCancelled(ptm.k) {
+	// If the request was cancelled, make sure we clean up the request tracker
+	if ok && data.lt.WasCancelled(ptm.k) {
+		data.lt.RemoveRequest(ptm.k)
```

Comment:
@hannahhoward can you remember why we don't just remove the request on cancel?

Reply:
Ah, nevermind. @dirkmc pointed me to https://github.com/ipfs/go-bitswap/blob/5204f405908426e1d7951e2632a76a0b9a82fff2/sessionpeermanager/sessionpeermanager_test.go#L335.

So, I _think_ the correct fix is to detect that all peers have sent back their responses. _This_ fix will only record one latency.

---

However, this _shouldn't_ be an issue as we should be hitting the timeout and removing all these trackers anyways. Something else is fishy.

Reply:
I think the latency tracker records latency per-peer.

Each [peerData](https://github.com/ipfs/go-bitswap/blob/master/sessionpeermanager/peerdata.go#L13) has an associated latency tracker:
```go
type peerData struct {
	hasLatency bool
	latency    time.Duration
	lt         *latencyTracker
}
```

The [latency tracker](https://github.com/ipfs/go-bitswap/blob/master/sessionpeermanager/latencytracker.go#L9) keeps track of the time that a request for a CID was sent to a peer, whether a cancel was then issued, and how long it took for a response to arrive:

```go
type requestData struct {
	startedAt    time.Time
	wasCancelled bool
	timeoutFunc  *time.Timer
}

type latencyTracker struct {
	requests map[cid.Cid]*requestData
}
```

There are broadcast requests, and requests to a specific set of peers
- if the request was a broadcast, then if the request times out we [clean up the latency tracker data](https://github.com/ipfs/go-bitswap/blob/master/sessionpeermanager/sessionpeermanager.go#L341) for the request for that CID
- if the request was to a specific peer, then if the request times out we [call `recordResponse()`](https://github.com/ipfs/go-bitswap/blob/master/sessionpeermanager/sessionpeermanager.go#L332). [`recordResponse()`](https://github.com/ipfs/go-bitswap/blob/master/sessionpeermanager/sessionpeermanager.go#L201) [calls `data.AdjustLatency()`](https://github.com/ipfs/go-bitswap/blob/master/sessionpeermanager/sessionpeermanager.go#L215) on the `peerData`, which in turn [cleans up the latency tracker data](https://github.com/ipfs/go-bitswap/blob/master/sessionpeermanager/peerdata.go#L29) for the request for that CID.
The problem is that if the request was cancelled, we don't call `recordResponse()` so the latency tracker data for that request never gets cleaned up.

I think your initial suggestion makes the most sense - I don't know if it's worth the complexity and extra memory required to wait for a timeout after a cancel. I think we should just clean up  the latency tracker information as soon as there is a cancel.

Reply:
Ah, I missed the per-peer part. This code looks correct.

> I think we should just clean up the latency tracker information as soon as there is a cancel.

Without tracking latencies for canceled wants, we may not correctly update latencies. For example:

Assume peers 1 and 2 have recorded latencies of 1s.

1. Ask peers 1 and 2.
2. 2 seconds pass.
3. Get a response for peer 1.
4. Record a latency of 2s for peer 1.
5. Cancel the request to peer 2.
6. Remove the latency tracker.

We've just penalized peer 1 for responding before peer 2.

On the other hand, the _current_ code will penalize peer 1 and not peer 2 if peer 2 processes the cancel before sending the block.

The correct solution may be randomization. That is:

1. Record the _first_ latency and remove all other trackers.
2. Make sure to randomly sample peers that _aren't_ the fastest.

Reply:
Ah yes you're right, that's a tricky problem to solve 🤔